flang/lib/Lower/ReductionRuntime.cpp

```diff
@@ -0,0 +1,127 @@
+//===-- ReductionRuntime.cpp -- runtime for reduction intrinsics -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Lower/ReductionRuntime.h"
+#include "../../runtime/reduction.h"
+#include "RTBuilder.h"
+#include "flang/Lower/Bridge.h"
+#include "flang/Lower/CharacterExpr.h"
+#include "flang/Lower/FIRBuilder.h"
+#include "flang/Lower/Support/BoxValue.h"
+#include "flang/Lower/Todo.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+
+using namespace Fortran::runtime;
+
+/// Generate calls to reduction intrinsics such as All and Any.
+/// These are the descriptor based implementations that take two 
+/// arguments (mask, dim).
+template <typename FN>
+static void
+genRed2Args(FN func, Fortran::lower::FirOpBuilder &builder,
+            mlir::Location loc, mlir::Value resultBox,
+            mlir::Value maskBox, mlir::Value dim);
```

Comment:
It doesn't look like you need to forward declare this template but can just move the definition to here.